courseScheduleII.js

```diff
@@ -0,0 +1,113 @@
+/*
+For this approach, we can represent the input as a dependency graph where in order to take a course, 
+you have to complete a prereq, which is the same as two nodes connected by an edge.
+
+Since we are trying to find a valid ordering for all these courses that respects our prereqs,
+we can approach this problem with a topological sort algorithm where if there happens to be an
+edge between a prereq and a class, it means that the prereq appears before the class in a linear ordering fashion.
+
+A class that has a prereq before would be an indegree relationship and we can the can use our topological sort algorithm to track
+the number of indegrees. We can create a hashmap, in this case a JavsScript object, that maps all the classes
+to its prereqs and decrease the number of indegrees once we process that class.
+
+That way we're left with 0 indegree classes in a valid order.
```

Comment:
I'm going to do my best to explain how this algorithm works, and hopefully that can improve your intuition.

Let's suppose you were a student at a university, and you want to complete all the required computer science courses at your university. You can model the prerequisite structure as a dependency graph.

Intuitively, how would you approach completing all these courses while meeting the prereqs? Well, you'd do the obvious thing. Find all the courses that have 0 prereqs, and take them during your first semester. Then, during your second semester, find all the classes that either have 0 prereqs, or exclusively have prereqs that you have taken. Take these classes. Next semester, repeat this process.

The idea is that at the start of every semester, you ask "which classes am I now eligible to take" and you take those. In the beginning, that will just be the classes with no prereqs (for example, an introductory programming course). Over time, the set of classes that you're eligible to take will grow larger.

This is pretty much what the BFS topological sort algorithm does (although it is NOT what the DFS topological sort algorithm does). You start by finding nodes with indegree 0. In the context of classes, this is the same as finding all of the classes that have NO prereqs. So, you may notice that indegree is the same thing is as number of prereqs. Then, you "take" all those indegree 0 classes by processing them in your queue. The key detail is that when you "take" a class, you delete all **outgoing** edges from that node. Intuitively, this corresponds to saying "oh, I no longer need to worry about this prerequisite".

Let's consider a concrete example. Suppose we have the following 3 courses:
CS1 - prereqs: none
CS2 - prereqs: CS1
CS3 - prereqs: CS1, CS2

Now you take CS1. Mentally, you can say "oh okay, I no longer need to worry about the prereq of CS1 anywhere." So, you can update your mental model to look like this:
CS2 - prereqs: ~~CS1~~
CS3 - prereqs: ~~CS1~~, CS2
(or you can just delete the CS1 prereqs, which is what our algorithm does).

And you'd repeat this process every time you finish a class.

So, to summarize, your algorithm is just doing the obvious thing that a person would do! Find all the classes with no prereqs and take them. Now, mentally "delete" all the prereqs that you have taken care of, and your procedure. Do this until you've taken all the classes.

I hope this clarified things, and now you can try writing this explanation again. If you're confused, feel free to message me.

Reply:
Thanks! I believe it does. Here's my explanation:

Basically, our algorithm works by first taking all the indegree 0 classes (classes/prereqs that do not have any incoming edges) and adds them to our queue. And for every class that is processed, is it pushed to our valid sorting array. If a particular class has an edge with another class (prereq for a class), we grab that class and decrease the number on indegree by 1 until it's 0 and then added to queue.

Reply:
Yup, that's right!